akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -514,7 +514,29 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
   /**
    * Requests to stop receiving events from a given input port. Cancelling clears any ungrabbed elements from the port.
    */
-  final protected def cancel[T](in: Inlet[T], cause: Throwable): Unit = interpreter.cancel(conn(in), cause)
+  final protected def cancel[T](in: Inlet[T], cause: Throwable): Unit = cancel(conn(in), cause)
+
+  private[this] lazy val _delayedCancellationCallback = getAsyncCallback[(Connection, Throwable)] {
+    case (connection, cause) =>
+      interpreter.log.debug(s"Now cancelling $connection in stage $this")
+      doCancel(connection, cause)
+  }
+  private def cancel[T](connection: Connection, cause: Throwable): Unit =
+    attributes.mandatoryAttribute[Attributes.CancellationStrategy].strategy match {
+      case Attributes.CancellationStrategy.AfterDelay(delay, _) =>
+        // since the port is not actually cancelled, we install a handler to ignore upcoming elements
+        connection.inHandler = new InHandler {
+          // ignore pushs now, since the stage wanted it cancelled already
+          override def onPush(): Unit = ()
+          // do not ignore termination signals
+        }
```

Comment:
If I look at `setHandler` it does two assignments, does this need to update the `handlers` array as well?

Reply:
Good point, I think the problem is that it is somewhat hard to find the right idx into `handlers`. There might be another problem which is that the stage may choose to select another `InHandler` after cancellation and then observe an element it shouldn't observe. But maybe we can document ourselves out of that situation?

Reply:
Or make this one a constant and not allow switching away from it with `setHandler`?

Couldn't we just take the previous connection.handler and indexOf it in handlers?

Reply:
The stage could use the same handler for multiple inputs so that doesn't work but it probably doesn't matter much if we prevent setting new handlers. This input is done anyway after `cancel` has been called. Whatever we do here would only be for additional diagnostics in cases of user error.